toffy/fov_watcher.py

```diff
@@ -0,0 +1,217 @@
+        """
+
+        # TODO: check watchdog path depth
+
+        fov_name, extension = path.split('.')[0:2]
+        if fov_name in self.completion:
+            if extension in self.completion[fov_name]:
+                self.completion[fov_name][extension] = True
+
+            if all(self.completion[fov_name].values):
+                return True, fov_name
+
+        return False, fov_name
+
+    def check_completion(self) -> dict:
+        """Condenses internal dictionary to show which fovs have finished
+
+        Returns:
+            dict
+        """
+        return {k: all(self.completion[k].values) for k in self.completion}
+
+
+class FOV_EventHandler(FileSystemEventHandler):
+    """File event handler for FOV files
+
+    Attributes:
+        run_folder (str):
+            path to run folder
+        watcher_out (str):
+            folder to save all callback results + log file
+        run_structure (RunStructure):
+            expected run file structure + completion status
+        per_fov (list):
+            callbacks to run on each fov
+        per_run (list):
+            callbacks to run over the entire run
+        panel (tuple | pd.DataFrame):
+            masses to extract
+    """
+    def __init__(self, run_folder: str, per_fov: List[Callable[[xr.DataArray, str], None]],
+                 per_run: List[Callable[[str, str], None]],
+                 panel: Union[Tuple, pd.DataFrame] = (0.3, 0.0)):
+        """Initializes FOV_EventHandler
+
+        Args:
+            run_folder (str):
+                path to run folder
+            per_fov (list):
+                callbacks to run on each fov
+            per_run (list):
+                callbacks to run over the entire run
+            panel (tuple | pd.DataFrame):
+                masses to extract
+        """
+        super().__init__()
+        self.run_folder = run_folder
+
+        self.watcher_out = os.path.join(run_folder, 'watcher_outs')
+
+        if not os.path.exists(self.watcher_out):
+            os.makedirs(self.watcher_out)
+
+        # create run structure
+        self.run_structure = RunStructure(run_folder)
+
+        self.per_fov = per_fov
+        self.per_run = per_run
+        self.panel = panel
+
+        for root, dirs, files in os.walk(run_folder):
+            for name in files:
+                self.on_created(FileCreatedEvent(os.path.join(root, name)))
+
+    def on_created(self, event: FileCreatedEvent):
+        """Handles file creation events
+
+        If FOV structure is completed, all callbacks in `per_fov` will be run over the data.
+        This function is automatically called; users generally shouldn't call this function
+
+        Args:
+            event (FileCreatedEvent):
+                file creation event
+        """
+        super().on_created(event)
+
+        # check if what's created is in the run structure
+        fov_ready, point_name = self.run_structure.check_run_condition(event.src_path)
```

Comment:
I think this will need to be modified to account for the fact that files are created with size zero, and then later copied over.

Reply:
for sure; I think I can just set a small delay (like 10s?) to wait for file download completion. since this already runs asynchronously, a delay shouldn't interfere with the file watching or further queued callbacks.

Reply:
Sometimes it takes seconds, sometimes it's like 3 minutes. There's never an in between; it's either size zero or completely copied over. Can you check that the file has a non-zeo size?

Reply:
oh awesome, yeah that's easy. I might add a timeout argument (default like 10mins or something), just so things don't hang if something never arrives

Reply:
Yup good idea, sometime the first one will never copy over